kivy/properties.pyx

```diff
@@ -387,15 +387,55 @@ cdef class Property:
         return '<{} name={}>'.format(self.__class__.__name__, self._name)
 
     cdef init_storage(self, EventDispatcher obj, PropertyStorage storage):
-        storage.value = self.convert(obj, self.defaultvalue)
+        storage.value = self.convert(obj, self.defaultvalue, storage)
         storage.observers = EventObservers.__new__(EventObservers)
         storage.property_obj = self
 
     cdef PropertyStorage create_property_storage(self):
         """Returns a new property storage used by this property."""
         return PropertyStorage.__new__(PropertyStorage)
 
-    cpdef link(self, EventDispatcher obj, str name):
+    cdef PropertyStorage get_property_storage(self, EventDispatcher obj):
+        cdef PropertyStorage ps = obj.__storage.get(self._name)
+        if ps is None:
+            self.link(obj, self._name)
+            self.link_deps(obj, self._name)
+            ps = obj.__storage[self._name]
+        return ps
+
+    def __set_name__(self, owner, name):
+        if name == 'touch_down' or name == 'touch_move' or name == 'touch_up':
```

Comment:
Can we move this `touch_xxx` and other event name tests in `register_event_type` method? Then `register_event_type` can check if property name or event name already exists and raise an error?

Reply:
While I understand what you mean and it does seem like a better idea, there are multiple issues with it. Currently we only ensure that someone doesn't create a *kivy* property that would have a event handler with the same name. But we don't and shouldn't stop people from creating properties with that name. I.e.:

```py
class A(Widget):
    def __init__(self):
        self.touch_down =True
```

But perhaps that's not what you meant. But rather that when creating a Kivy prop we check whether an event with that name exists and raise an exception, and similarly raise an exception when creating an event when a property with that name already exists. Currently we don't do it, except for these few special names so doing it now could easily break existing code.

Perhaps we should only do it when registering these automatic `on_xxx` handlers when the instance is created, because that's the only reason it's a problem (and perhaps for kv `on_xxx: ` handlers) because of the ambiguity. But that couldn't be part of this PR and would need a discussion to figure out if the benefits are worth the potential breakage, and I'm not sure about it.

Reply:
I meant we check if event with property name already exists before registering the event and check if event name exist before registering the property. If any exists then raise the error.

Reply:
This would have to happen in EventDispatcher init as we register events, because this code is executed before any events are registered. But we could raise an exception if an event is registered for an existing property.

Similarly, we could also raise an exception if KV or user manually create a property and an event with the name exists.

But that would need a PR of its own with tests etc. And, the bigger issue, that would be a breaking change. Perhaps it's not a problem given that if both a property and event with same name exists the user code may not function as they expected. But it would need discussion to see if the break is worth it. I'm on the fence.